### ISSUE TITLE
Seems datachannel.send() blocked by reading channel

### ISSUE THREAD

**Issue by alex-eri** [jlaine/aiortc#40]
`while True` in rtcsctptransport.py blocks event loop if queue full [on receive here](https://github.com/jlaine/aiortc/blob/5e21a99235c4df37bdb3fa128689cfa668554645/aiortc/rtcsctptransport.py#L488) 

My suggestion is to rewrite it using `asyncio.Queue` but it is not trivial

**Comment by alex-eri** [comment]
`while True` in rtcsctptransport.py blocks event loop if queue full [on receive here](https://github.com/jlaine/aiortc/blob/5e21a99235c4df37bdb3fa128689cfa668554645/aiortc/rtcsctptransport.py#L488) and [on transmit here](https://github.com/jlaine/aiortc/blob/5e21a99235c4df37bdb3fa128689cfa668554645/aiortc/rtcsctptransport.py#L1139)


My suggestion is to rewrite it using `asyncio.Queue`.

**Comment by alex-eri** [comment]
Can we replace _data_channel_flush with task?

**Comment by jlaine** [comment]
I'm not very happy with the current DataChannel / SCTP boundary so it will almost certainly need refactoring. As there are very few userspace implementations out there (actually usrsctp is the only one I found) I haven't had many examples to draw inspiration from.

My main problem is that RTCDataChannel.send pushes data down to the SCTP layer without any kind of backpressure. This is particularly visible in the filexfer example: you will end up reading the whole file and shoving  into memory which is obviously undesirable for large files. A review of how browsers handle buffering would be most welcome

**Comment by alex-eri** [comment]
tried with queues with no success.

I thinking about unbuffer channels

What about `socket.socketpair` ?

udp - [ice+stun] - socketpair - [ sctp <= split\reorder ] - socketpair - client

So we connect [io.BufferedRandom](https://docs.python.org/3.3/library/io.html#io.BufferedRandom) to socketpair on client side for streaming big data. Small messages (by design) sended as is.

It writes a buffers to socket, some data buffered in kernel(unix socket) then blocks.
splitter reads socket with async selector, split buffer if nessesary and put it to ICE socket.

ICE forward it to UDP.

On other side buffer still needed to reorder packets.

**Comment by alex-eri** [comment]
Or connect ice to sctp directlly without queue - callback based api.

**Comment by jlaine** [comment]
I have to admit I didn't fully understand your issue, what's blocking exactly? The only "while True" in rtcsctptransport.py is the consumer for received data, which doesn't call datachannel.send().

**Comment by jlaine** [comment]
I'm a bit surprised, as the `while True` reader loop definitely does have an `await`:

https://github.com/jlaine/aiortc/blob/5e21a99235c4df37bdb3fa128689cfa668554645/aiortc/rtcsctptransport.py#L681

**Comment by nickaknudson** [comment]
I'm also seeing symptoms of this in `0.9.0` generating the following error:

`RuntimeWarning: coroutine 'RTCDataChannel.send' was never awaited`

Basically unable to send anything over a data channel. Reverting to `0.8.0` resolves the issue.

**Comment by jlaine** [comment]
@nickaknudson that's unrelated: as mentioned in the changelog send() is now a coroutine so you need to do:

await channel.send("Foo")

**Comment by jlaine** [comment]
After discussion with @lgrahl I've reverted the change which made `RTCDataChannel.send()` a coroutine. 

I'm pushing 0.9.1 out to restore pre-0.9.0 behaviour, sorry for the confusion.

**Comment by jlaine** [comment]
@alex-eri I'm going to close this issue, as there is now a cap on how much data gets transfered from RTCDataChannel to the SCTP layer.

**Comment by lgrahl** [comment]
The need to control the flow is only required for SCTP since DTLS/UDP has no flow control. I would propose to add a `drain` method. You can take the reader/writer classes of asyncio as inspiration.

**Comment by lgrahl** [comment]
Something like this:

```python
class DataChannel:
    def __init__():
        self._send_queue = asyncio.Queue(maxsize=1)
        self._receive_queue = asyncio.Queue(maxsize=1)

    async def send(message: Union[bytes, asyncio.StreamWriter]):
        if isinstance(message, bytes):
            message = bytes_to_stream(message)
        return await self._send_queue.put(message)

    async def receive(as_bytes: bool = False) -> Union[bytes, asyncio.StreamReader]:
        message = await self._receive_queue.get()
        if as_bytes:
            message = await stream_to_bytes(message)
        return message
```

If the user is interested in controlling the flow in its fullest form, it needs to use streams for messages. Otherwise, if it expects message size to not be an issue, it can also utilise bytes.

**Comment by lgrahl** [comment]
^ FYI: I might actually pick this up at some point and implement it myself. Don't rely on it though. :smiley_cat:

**Comment by jlaine** [comment]
Cool! Could you maybe open a separate issue to track it (with your code sample)?
